sdk/python/test_data/pipelines/pipeline_with_parallelfor_list_artifacts.py

```diff
@@ -0,0 +1,77 @@
+# Copyright 2022 The Kubeflow Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import List
+
+from kfp import compiler
+from kfp import dsl
+from kfp.dsl import Artifact
+from kfp.dsl import Dataset
+
+
+@dsl.component
+def print_artifact_name(artifact: Artifact) -> str:
+    print(artifact.name)
+    return artifact.name
+
+
+@dsl.component
+def make_dataset(text: str) -> Dataset:
+    dataset = Dataset(uri=dsl.get_uri(), metadata={'length': len(text)})
+    with open(dataset.path, 'w') as f:
+        f.write(text)
+    return dataset
+
+
+@dsl.pipeline
+def make_datasets(
+        texts: List[str] = ['Hello', ',', ' ', 'world!']) -> List[Dataset]:
+    with dsl.ParallelFor(texts) as text:
+        t1 = make_dataset(text=text)
+
+    return dsl.Collected(t1.output)
+
+
+@dsl.component
+def make_artifact(name: str) -> Artifact:
+    artifact = Artifact(uri=dsl.get_uri(), metadata={'length': len(name)})
+    with open(artifact.path, 'w') as f:
+        f.write(name)
+    return artifact
+
+
+@dsl.pipeline
+def make_artifacts(
+        names: List[str] = ['Hello', ',', ' ', 'world!']) -> List[Artifact]:
+    with dsl.ParallelFor(names) as name:
+        t1 = make_artifact(name=name)
+
+    return dsl.Collected(t1.output)
+
```

Comment:
I think it would be good to test some of the other pipeline topologies that a user may express. For example, I believe this should run but does not:

```python
@dsl.pipeline
def my_pipeline(
        names: List[str] = ['Hello', ',', ' ', 'world!']) -> List[Artifact]:
    with dsl.ParallelFor(names) as name:
        t1 = make_artifact(name=name)

    with dsl.ParallelFor(dsl.Collected(t1.output)) as item:
        print_artifact_name(artifact=item)
```

We can do so in this file. It also may be helpful to do this in `compiler_test.py`, that way we can split out the test cases into clearly named methods to ensure we're capturing a representative set of topologies.

Reply:
I tried out the corresponding code for parameters in a clean venv (installed kfp from pip) and I also got an error
```
@dsl.component
def str_identity(s: str) -> str:
    return s

@dsl.pipeline
def my_pipeline(): 
  texts = ['Hello', ',', ' ', 'world!']
  with dsl.ParallelFor(texts) as text:
      t1 = str_identity(s=text)

  with dsl.ParallelFor(dsl.Collected(t1.output)) as item:
      str_identity(s=item)
```

Seems like this issue affects parameters as well as artifacts, wdyt about resolving this in the next pr?

Reply:
Good catch! Addressing as a follow-up sounds good.

Reply:
Good catch. Fixing in a follow-up PR sounds good!